pymongo/network_layer.py

```diff
@@ -413,3 +379,467 @@ def receive_data(conn: Connection, length: int, deadline: Optional[float]) -> me
+        return self.conn[0].get_extra_info("socket")
+
+
+class NetworkingInterface(NetworkingInterfaceBase):
+    def __init__(self, conn: Union[socket.socket, _sslConn]):
+        super().__init__(conn)
+
+    def gettimeout(self) -> float | None:
+        return self.conn.gettimeout()
+
+    def settimeout(self, timeout: float | None) -> None:
+        self.conn.settimeout(timeout)
+
+    def close(self) -> None:
+        self.conn.close()
+
+    def is_closing(self) -> bool:
+        return self.conn.is_closing()
+
+    @property
+    def get_conn(self) -> Union[socket.socket, _sslConn]:
+        return self.conn
+
+    @property
+    def sock(self) -> Union[socket.socket, _sslConn]:
+        return self.conn
+
+    def fileno(self) -> int:
+        return self.conn.fileno()
+
+    def recv_into(self, buffer: bytes) -> int:
+        return self.conn.recv_into(buffer)
+
+
+class PyMongoProtocol(BufferedProtocol):
+    def __init__(self, timeout: Optional[float] = None, buffer_size: int = 2**14):
+        self._buffer_size = buffer_size
+        self.transport: Transport = None  # type: ignore[assignment]
+        self._buffer = memoryview(bytearray(self._buffer_size))
+        self._overflow: Optional[memoryview] = None
+        self._start_index = 0
+        self._end_index = 0
+        self._overflow_index = 0
+        self._body_size = 0
+        self._op_code: int = None  # type: ignore[assignment]
+        self._connection_lost = False
+        self._paused = False
+        self._drain_waiter: Optional[Future] = None
+        self._read_waiter: Optional[Future] = None
+        self._timeout = timeout
+        self._is_compressed = False
+        self._compressor_id: Optional[int] = None
+        self._max_message_size = MAX_MESSAGE_SIZE
+        self._request_id: Optional[int] = None
+        self._closed = asyncio.get_running_loop().create_future()
+        self._expecting_header = True
+        self._pending_messages: collections.deque[Future] = collections.deque()
+        self._done_messages: collections.deque[Future] = collections.deque()
+
+    def settimeout(self, timeout: float | None) -> None:
+        self._timeout = timeout
+
+    @property
+    def gettimeout(self) -> float | None:
+        """The configured timeout for the socket that underlies our protocol pair."""
+        return self._timeout
+
+    def connection_made(self, transport: BaseTransport) -> None:
+        """Called exactly once when a connection is made.
+        The transport argument is the transport representing the write side of the connection.
+        """
+        self.transport = transport  # type: ignore[assignment]
+
+    async def write(self, message: bytes) -> None:
+        """Write a message to this connection's transport."""
+        if self.transport.is_closing():
+            raise OSError("Connection is closed")
+        self.transport.write(message)
+        await self._drain_helper()
+        self.transport.resume_reading()
```

Comment:
What if we resume _before_ writing?:
```python
        self.transport.resume_reading()
        self.transport.write(message)
```
Will that improve latency?

We can answer this in a follow-up ticket.